{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "require([\"codemirror/keymap/sublime\", \"notebook/js/cell\", \"base/js/namespace\"],\n",
       "    function(sublime_keymap, cell, IPython) {\n",
       "        cell.Cell.options_default.cm_config.keyMap = 'sublime';\n",
       "    \n",
       "        var cells = IPython.notebook.get_cells();\n",
       "        for(var cl=0; cl< cells.length ; cl++){\n",
       "            cells[cl].code_mirror.setOption('keyMap', 'sublime');\n",
       "        }\n",
       "    }\n",
       ");\n"
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%javascript\n",
    "require([\"codemirror/keymap/sublime\", \"notebook/js/cell\", \"base/js/namespace\"],\n",
    "    function(sublime_keymap, cell, IPython) {\n",
    "        cell.Cell.options_default.cm_config.keyMap = 'sublime';\n",
    "    \n",
    "        var cells = IPython.notebook.get_cells();\n",
    "        for(var cl=0; cl< cells.length ; cl++){\n",
    "            cells[cl].code_mirror.setOption('keyMap', 'sublime');\n",
    "        }\n",
    "    }\n",
    ");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width: 90% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# change the cell width\n",
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.container { width: 90% !important; }</style>\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "import os\n",
    "import cooler\n",
    "import cooltools\n",
    "import numpy as np\n",
    "from cooltools import eigdecomp\n",
    "import bioframe\n",
    "from pathlib import Path\n",
    "import multiprocess as mp\n",
    "\n",
    "from copy import copy\n",
    "\n",
    "%matplotlib inline\n",
    "%config InlineBackend.print_figure_kwargs={'facecolor' : \"w\"}\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import LogNorm\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### some of the input cooler files that we used to combine/recombine/subset etc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "ins_samples = {\n",
    "    \"Hap1-WT-combined.hg19\" : \"Hap1-WT-combined.mcool\",\n",
    "    \"PolII-IAA.hg19\" : \"PolII-IAA.hg19.mapq_30.1000.mcool\",\n",
    "    \"PolII-NT.hg19\" : \"PolII-NT.hg19.mapq_30.1000.mcool\",\n",
    "    \"all_CTCF_combined.hg19\" : \"all_combined.mcool\",\n",
    "    \"noCTCF_IAA_combined.hg19\" : \"noCTCF_IAA_combined.mcool\",\n",
    "    \n",
    "    \"all_CTCF_combined.hg19\" : \"all_combined.mcool\",\n",
    "    \"noCTCF_IAA_combined.hg19\" : \"noCTCF_IAA_combined.mcool\",\n",
    "    \n",
    "    \"CkoCT442_NT_pool.hg19\" : \"CkoCT442-NT-pool.mcool\",\n",
    "    \"CkoCT442_IAA_pool.hg19\" : \"CkoCT442-IAA-pool.mcool\",\n",
    "    # RAD21 degron for HCT from Rao et al 2017 ...\n",
    "    \"RAD21-IAA.hg19\" : \"RAD21-AID-IAA-6H.hg19.mapq_30.1000.mcool\",\n",
    "    \"RAD21-NT.hg19\" : \"RAD21-AID-NT.hg19.mapq_30.1000.mcool\",\n",
    "\n",
    "    \"CkoC442-IAA48H-R1-T1.hg19\" : \"CkoC442-IAA48H-R1-T1__hg19.hg19.mapq_30.1000.mcool\",\n",
    "    \"CkoC442-NT-R1-T1.hg19\" : \"./wt_like_coolers_to_combine/CkoC442-NT-R1-T1__hg19.hg19.mapq_30.1000.mcool\",\n",
    "    \"CkoCT442-AAVS1sg2-4-NT-R1-T1.hg19\" : \"CkoCT442-AAVS1sg2-4-NT-R1-T1__hg19.hg19.mapq_30.1000.mcool\",\n",
    "    \"CkoCT442-DDX55sg2-B-NT-R1-T1.hg19\" : \"CkoCT442-DDX55sg2-B-NT-R1-T1__hg19.hg19.mapq_30.1000.mcool\",\n",
    "    \"G1-NT.hg19\" : \"G1-NT.hg19.1kb.mcool\",\n",
    "    \"G1-IAA.hg19\" : \"G1-IAA.hg19.1kb.mcool\",\n",
    "    \n",
    "    # Mutants time !!!\n",
    "    # Knock-outs - genes were deleted ...\n",
    "    \"mutControl-NT.hg19\":\"CkoCT442-AAVS1-NT-pool.mcool\",\n",
    "    \"mutControl-IAA.hg19\":\"CkoCT442-AAVS1-IAA-pool.mcool\",\n",
    "    \"mutDDX55-NT.hg19\":\"DDX55-clones-NT.hg19.mapq_30.1000.mcool\",\n",
    "    \"mutDDX55-IAA.hg19\":\"DDX55-clones-IAA.hg19.mapq_30.1000.mcool\",\n",
    "    \"mutTAF5L-NT.hg19\":\"TAF5L-clones-NT.hg19.mapq_30.1000.mcool\",\n",
    "    \"mutTAF5L-IAA.hg19\":\"TAF5L-clones-IAA.hg19.mapq_30.1000.mcool\",\n",
    "    # Knock-downs - genes were silenced/deplted (siRNA) ...\n",
    "    \"siControl-NT.hg19\":\"siCTRL-NT.hg19.mapq_30.1000.mcool\",\n",
    "    \"siControl-IAA.hg19\":\"siCTRL-IAA.hg19.mapq_30.1000.mcool\",\n",
    "    \"siDDX55-NT.hg19\":\"siDDX55-NT.hg19.mapq_30.1000.mcool\",\n",
    "    \"siDDX55-IAA.hg19\":\"siDDX55-IAA.hg19.mapq_30.1000.mcool\",\n",
    "    \"siTAF5L-NT.hg19\":\"siTAF5L-NT.hg19.mapq_30.1000.mcool\",\n",
    "    \"siTAF5L-IAA.hg19\":\"siTAF5L-IAA.hg19.mapq_30.1000.mcool\",\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%bash\n",
    "# cname=\"./CkoCT442-NT-pool.mcool\"\n",
    "# for uri in $(cooler ls $cname);\n",
    "# do\n",
    "#     echo $uri\n",
    "#     cooler balance --check $uri\n",
    "# done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "already exist !\n",
      "Hap1-WT-combined.hg19.5kb.100kb.bw\n",
      "already exist !\n",
      "PolII-IAA.hg19.5kb.100kb.bw\n",
      "already exist !\n",
      "PolII-NT.hg19.5kb.100kb.bw\n",
      "already exist !\n",
      "all_CTCF_combined.hg19.5kb.100kb.bw\n",
      "already exist !\n",
      "noCTCF_IAA_combined.hg19.5kb.100kb.bw\n",
      "already exist !\n",
      "CkoCT442_NT_pool.hg19.5kb.100kb.bw\n",
      "already exist !\n",
      "CkoCT442_IAA_pool.hg19.5kb.100kb.bw\n",
      "already exist !\n",
      "RAD21-IAA.hg19.5kb.100kb.bw\n",
      "already exist !\n",
      "RAD21-NT.hg19.5kb.100kb.bw\n",
      "already exist !\n",
      "CkoC442-IAA48H-R1-T1.hg19.5kb.100kb.bw\n",
      "already exist !\n",
      "CkoC442-NT-R1-T1.hg19.5kb.100kb.bw\n",
      "already exist !\n",
      "CkoCT442-AAVS1sg2-4-NT-R1-T1.hg19.5kb.100kb.bw\n",
      "already exist !\n",
      "CkoCT442-DDX55sg2-B-NT-R1-T1.hg19.5kb.100kb.bw\n",
      "already exist !\n",
      "G1-NT.hg19.5kb.100kb.bw\n",
      "already exist !\n",
      "G1-IAA.hg19.5kb.100kb.bw\n",
      "working on ...\n",
      "mutControl-NT.hg19 CkoCT442-AAVS1-NT-pool.mcool\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n",
      "bedGraphToBigWig /tmp/tmpfgzii_rd.bg /tmp/tmpdcg_w3ks.chrom.sizes mutControl-NT.hg19.5kb.100kb.bw\n",
      "working on ...\n",
      "mutControl-IAA.hg19 CkoCT442-AAVS1-IAA-pool.mcool\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n",
      "bedGraphToBigWig /tmp/tmpatrxlwlj.bg /tmp/tmpxaqqgkba.chrom.sizes mutControl-IAA.hg19.5kb.100kb.bw\n",
      "working on ...\n",
      "mutDDX55-NT.hg19 DDX55-clones-NT.hg19.mapq_30.1000.mcool\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n",
      "bedGraphToBigWig /tmp/tmps4613nkr.bg /tmp/tmp9gn62z1k.chrom.sizes mutDDX55-NT.hg19.5kb.100kb.bw\n",
      "working on ...\n",
      "mutDDX55-IAA.hg19 DDX55-clones-IAA.hg19.mapq_30.1000.mcool\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n",
      "bedGraphToBigWig /tmp/tmp2xuiojse.bg /tmp/tmpieo3xwo0.chrom.sizes mutDDX55-IAA.hg19.5kb.100kb.bw\n",
      "working on ...\n",
      "mutTAF5L-NT.hg19 TAF5L-clones-NT.hg19.mapq_30.1000.mcool\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n",
      "bedGraphToBigWig /tmp/tmpcymd4yf6.bg /tmp/tmpeq110w6d.chrom.sizes mutTAF5L-NT.hg19.5kb.100kb.bw\n",
      "working on ...\n",
      "mutTAF5L-IAA.hg19 TAF5L-clones-IAA.hg19.mapq_30.1000.mcool\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n",
      "bedGraphToBigWig /tmp/tmp64j6q35n.bg /tmp/tmp2_x961r6.chrom.sizes mutTAF5L-IAA.hg19.5kb.100kb.bw\n",
      "working on ...\n",
      "siControl-NT.hg19 siCTRL-NT.hg19.mapq_30.1000.mcool\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n",
      "bedGraphToBigWig /tmp/tmp0uj7zzdj.bg /tmp/tmpkz7lsb17.chrom.sizes siControl-NT.hg19.5kb.100kb.bw\n",
      "working on ...\n",
      "siControl-IAA.hg19 siCTRL-IAA.hg19.mapq_30.1000.mcool\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n",
      "bedGraphToBigWig /tmp/tmp9ey3vydo.bg /tmp/tmplyvist0q.chrom.sizes siControl-IAA.hg19.5kb.100kb.bw\n",
      "working on ...\n",
      "siDDX55-NT.hg19 siDDX55-NT.hg19.mapq_30.1000.mcool\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n",
      "bedGraphToBigWig /tmp/tmpqrvrqu3h.bg /tmp/tmpozlhxe5t.chrom.sizes siDDX55-NT.hg19.5kb.100kb.bw\n",
      "working on ...\n",
      "siDDX55-IAA.hg19 siDDX55-IAA.hg19.mapq_30.1000.mcool\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n",
      "bedGraphToBigWig /tmp/tmpl1et2t_4.bg /tmp/tmphunqb61k.chrom.sizes siDDX55-IAA.hg19.5kb.100kb.bw\n",
      "working on ...\n",
      "siTAF5L-NT.hg19 siTAF5L-NT.hg19.mapq_30.1000.mcool\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n",
      "bedGraphToBigWig /tmp/tmpe_uzv6sa.bg /tmp/tmpo3xj4_lh.chrom.sizes siTAF5L-NT.hg19.5kb.100kb.bw\n",
      "working on ...\n",
      "siTAF5L-IAA.hg19 siTAF5L-IAA.hg19.mapq_30.1000.mcool\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n",
      "bedGraphToBigWig /tmp/tmp8p0affb7.bg /tmp/tmpowqo4iem.chrom.sizes siTAF5L-IAA.hg19.5kb.100kb.bw\n"
     ]
    }
   ],
   "source": [
    "binsize = 5_000\n",
    "binsize_human = f\"{int(binsize/1_000)}kb\"\n",
    "diamond = 100_000\n",
    "diamond_human = f\"{int(diamond/1_000)}kb\"\n",
    "\n",
    "hg19_chromsizes = bioframe.fetch_chromsizes(\"hg19\")\n",
    "\n",
    "# iterate over samples to calculate insulation on:\n",
    "for k,cname in ins_samples.items():\n",
    "    target_bw_file = Path(f\"{k}.{binsize_human}.{diamond_human}.bw\")\n",
    "    if target_bw_file.is_file():\n",
    "        print(\"already exist !\")\n",
    "        print(target_bw_file)\n",
    "        continue\n",
    "    else:\n",
    "        print(\"working on ...\")\n",
    "        print(k,cname)\n",
    "        ######################################\n",
    "        ! cooltools diamond-insulation \\\n",
    "            -o {k}.{binsize_human}.{diamond_human}.tsv \\\n",
    "            {cname}::/resolutions/{binsize} \\\n",
    "            {diamond}\n",
    "        ###################################### \n",
    "        df = bioframe.read_table(\n",
    "                f\"{k}.{binsize_human}.{diamond_human}.tsv\",\n",
    "                header=0\n",
    "        )\n",
    "        ######################################\n",
    "        bioframe.to_bigwig(\n",
    "            df,\n",
    "            hg19_chromsizes,\n",
    "            f\"{k}.{binsize_human}.{diamond_human}.bw\",\n",
    "            value_field = f\"log2_insulation_score_{diamond}\"\n",
    "        )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # checking sequencing depth of several samples ...\n",
    "# !cooler info all_combined.mcool::/resolutions/25000 | grep sum\n",
    "# !cooler info all_combined.mcool::/resolutions/25000 | grep nnz\n",
    "# !cooler info noCTCF_IAA_combined.mcool::/resolutions/25000 | grep sum\n",
    "# !cooler info noCTCF_IAA_combined.mcool::/resolutions/25000 | grep nnz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculating compartment status per arm for several samples ..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rename_arms(name):\n",
    "    \"\"\"\n",
    "    turn UCSC genomic region name\n",
    "    into a human readable name of chrom arm\n",
    "    \"\"\"\n",
    "    chrom,coords = name.split(\":\")\n",
    "    start,end = coords.split(\"-\")\n",
    "    if start == \"0\":\n",
    "        return chrom+\"p\"\n",
    "    else:\n",
    "        return chrom+\"q\"\n",
    "\n",
    "hg19_chromsizes = bioframe.fetch_chromsizes('hg19', as_bed=True)\n",
    "hg19_cens = bioframe.fetch_centromeres('hg19')\n",
    "hg19_arms = bioframe.split(hg19_chromsizes, hg19_cens, cols_points=['chrom', 'mid'])\n",
    "# # Select only chromosomes that are present in the cooler. \n",
    "# # This step is typically not required! we call it only because the test data are reduced. \n",
    "# hg19_chromsizes = hg19_chromsizes.set_index(\"chrom\").loc[hg19_chromsizes.chrom].reset_index() \n",
    "hg19_arms = hg19_arms.set_index(\"chrom\").loc[hg19_chromsizes.chrom.values[:-1]].reset_index()\n",
    "# call this to automaticly assign names to chromosomal arms:\n",
    "hg19_arms = bioframe.parse_regions(hg19_arms)\n",
    "hg19_arms[\"name\"] = hg19_arms[\"name\"].apply(rename_arms)\n",
    "hg19_arms.head()\n",
    "hg19_arms.to_csv(\"hg19_arms.bed\",sep=\"\\t\",header=False,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "already exist !\n",
      "frac_gene_coverage.25kb.bedGraph\n"
     ]
    }
   ],
   "source": [
    "# get some cooler to extract bin table - for reference track extraction ...\n",
    "binsize = 25_000\n",
    "human_binsize = f\"{int(binsize/1_000)}kb\"\n",
    "db = \"hg19\"\n",
    "# input samples here\n",
    "# make sure we do not regenerate the same reference track again ...\n",
    "target_ref_track = Path(f\"frac_gene_coverage.{human_binsize}.bedGraph\")\n",
    "if target_ref_track.is_file():\n",
    "    print(\"already exist !\")\n",
    "    print(target_ref_track)\n",
    "else:\n",
    "    print(f\"working on ... {human_binsize} gene coverage for {db} ...\")\n",
    "    ######################################\n",
    "    # extract gene coverage for a given bin-size ...\n",
    "    # we need SOME cooler to get a bin table with that binsize ...\n",
    "    s= 'CkoCT442_NT_pool.hg19'\n",
    "    fname = ins_samples[s]\n",
    "    clr = cooler.Cooler(f\"{fname}::/resolutions/{binsize}\")\n",
    "    # gene cov extraction !\n",
    "    gene_cov_df = bioframe.frac_gene_coverage(\n",
    "                clr.bins()[:],\n",
    "                db,\n",
    "                )\n",
    "    # save coverage track ...\n",
    "    gene_cov_df[[\"chrom\",\"start\",\"end\",\"coverage\"]] \\\n",
    "        .to_csv(\n",
    "            target_ref_track,\n",
    "            sep=\"\\t\",\n",
    "            header=False,\n",
    "            index=False\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": false,
    "slideshow": {
     "slide_type": ""
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "already exist !\n",
      "Hap1-WT-combined.hg19.EV.25kb.cis.bw\n",
      "already exist !\n",
      "PolII-IAA.hg19.EV.25kb.cis.bw\n",
      "already exist !\n",
      "PolII-NT.hg19.EV.25kb.cis.bw\n",
      "already exist !\n",
      "all_CTCF_combined.hg19.EV.25kb.cis.bw\n",
      "already exist !\n",
      "noCTCF_IAA_combined.hg19.EV.25kb.cis.bw\n",
      "already exist !\n",
      "CkoCT442_NT_pool.hg19.EV.25kb.cis.bw\n",
      "already exist !\n",
      "CkoCT442_IAA_pool.hg19.EV.25kb.cis.bw\n",
      "already exist !\n",
      "RAD21-IAA.hg19.EV.25kb.cis.bw\n",
      "already exist !\n",
      "RAD21-NT.hg19.EV.25kb.cis.bw\n",
      "already exist !\n",
      "CkoC442-IAA48H-R1-T1.hg19.EV.25kb.cis.bw\n",
      "already exist !\n",
      "CkoC442-NT-R1-T1.hg19.EV.25kb.cis.bw\n",
      "already exist !\n",
      "CkoCT442-AAVS1sg2-4-NT-R1-T1.hg19.EV.25kb.cis.bw\n",
      "already exist !\n",
      "CkoCT442-DDX55sg2-B-NT-R1-T1.hg19.EV.25kb.cis.bw\n",
      "already exist !\n",
      "G1-NT.hg19.EV.25kb.cis.bw\n",
      "already exist !\n",
      "G1-IAA.hg19.EV.25kb.cis.bw\n",
      "working on ...\n",
      "mutControl-NT.hg19 CkoCT442-AAVS1-NT-pool.mcool\n",
      "to generate mutControl-NT.hg19.EV.25kb.cis.bw ...\n",
      "Tokenization took: 32.37 ms\n",
      "Type conversion took: 19.66 ms\n",
      "Parser memory cleanup took: 0.00 ms\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n",
      "bedGraphToBigWig /tmp/tmpkja_e467.bg /tmp/tmp_esghpbs.chrom.sizes mutControl-NT.hg19.EV.25kb.cis.bw\n",
      "working on ...\n",
      "mutControl-IAA.hg19 CkoCT442-AAVS1-IAA-pool.mcool\n",
      "to generate mutControl-IAA.hg19.EV.25kb.cis.bw ...\n",
      "Tokenization took: 26.31 ms\n",
      "Type conversion took: 16.32 ms\n",
      "Parser memory cleanup took: 0.01 ms\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n",
      "bedGraphToBigWig /tmp/tmp19gq5bvy.bg /tmp/tmp5q1y5837.chrom.sizes mutControl-IAA.hg19.EV.25kb.cis.bw\n",
      "working on ...\n",
      "mutDDX55-NT.hg19 DDX55-clones-NT.hg19.mapq_30.1000.mcool\n",
      "to generate mutDDX55-NT.hg19.EV.25kb.cis.bw ...\n",
      "Tokenization took: 43.16 ms\n",
      "Type conversion took: 17.99 ms\n",
      "Parser memory cleanup took: 0.01 ms\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n",
      "bedGraphToBigWig /tmp/tmpoz6adyye.bg /tmp/tmpycdlv3qr.chrom.sizes mutDDX55-NT.hg19.EV.25kb.cis.bw\n",
      "working on ...\n",
      "mutDDX55-IAA.hg19 DDX55-clones-IAA.hg19.mapq_30.1000.mcool\n",
      "to generate mutDDX55-IAA.hg19.EV.25kb.cis.bw ...\n",
      "Tokenization took: 24.22 ms\n",
      "Type conversion took: 15.48 ms\n",
      "Parser memory cleanup took: 0.01 ms\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n",
      "bedGraphToBigWig /tmp/tmp1r4vi9rf.bg /tmp/tmphsp7ipkg.chrom.sizes mutDDX55-IAA.hg19.EV.25kb.cis.bw\n",
      "working on ...\n",
      "mutTAF5L-NT.hg19 TAF5L-clones-NT.hg19.mapq_30.1000.mcool\n",
      "to generate mutTAF5L-NT.hg19.EV.25kb.cis.bw ...\n",
      "Tokenization took: 26.49 ms\n",
      "Type conversion took: 10.98 ms\n",
      "Parser memory cleanup took: 0.00 ms\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n",
      "bedGraphToBigWig /tmp/tmp8oodn52f.bg /tmp/tmpom9ouvbs.chrom.sizes mutTAF5L-NT.hg19.EV.25kb.cis.bw\n",
      "working on ...\n",
      "mutTAF5L-IAA.hg19 TAF5L-clones-IAA.hg19.mapq_30.1000.mcool\n",
      "to generate mutTAF5L-IAA.hg19.EV.25kb.cis.bw ...\n",
      "Tokenization took: 23.51 ms\n",
      "Type conversion took: 10.62 ms\n",
      "Parser memory cleanup took: 0.01 ms\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n",
      "bedGraphToBigWig /tmp/tmpw1nyr40r.bg /tmp/tmp9_a9pip4.chrom.sizes mutTAF5L-IAA.hg19.EV.25kb.cis.bw\n",
      "working on ...\n",
      "siControl-NT.hg19 siCTRL-NT.hg19.mapq_30.1000.mcool\n",
      "to generate siControl-NT.hg19.EV.25kb.cis.bw ...\n",
      "Tokenization took: 32.64 ms\n",
      "Type conversion took: 23.42 ms\n",
      "Parser memory cleanup took: 0.01 ms\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n",
      "bedGraphToBigWig /tmp/tmplj0eivk1.bg /tmp/tmpjcdvy49u.chrom.sizes siControl-NT.hg19.EV.25kb.cis.bw\n",
      "working on ...\n",
      "siControl-IAA.hg19 siCTRL-IAA.hg19.mapq_30.1000.mcool\n",
      "to generate siControl-IAA.hg19.EV.25kb.cis.bw ...\n",
      "Tokenization took: 33.18 ms\n",
      "Type conversion took: 16.70 ms\n",
      "Parser memory cleanup took: 0.01 ms\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n",
      "bedGraphToBigWig /tmp/tmpch97f53r.bg /tmp/tmpfsh5l21m.chrom.sizes siControl-IAA.hg19.EV.25kb.cis.bw\n",
      "working on ...\n",
      "siDDX55-NT.hg19 siDDX55-NT.hg19.mapq_30.1000.mcool\n",
      "to generate siDDX55-NT.hg19.EV.25kb.cis.bw ...\n",
      "Tokenization took: 48.96 ms\n",
      "Type conversion took: 24.78 ms\n",
      "Parser memory cleanup took: 0.02 ms\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n",
      "bedGraphToBigWig /tmp/tmpxxws_hgu.bg /tmp/tmpfjl6n50v.chrom.sizes siDDX55-NT.hg19.EV.25kb.cis.bw\n",
      "working on ...\n",
      "siDDX55-IAA.hg19 siDDX55-IAA.hg19.mapq_30.1000.mcool\n",
      "to generate siDDX55-IAA.hg19.EV.25kb.cis.bw ...\n",
      "Tokenization took: 25.11 ms\n",
      "Type conversion took: 14.78 ms\n",
      "Parser memory cleanup took: 0.01 ms\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n",
      "bedGraphToBigWig /tmp/tmpu_6xfq9l.bg /tmp/tmp95x9skt2.chrom.sizes siDDX55-IAA.hg19.EV.25kb.cis.bw\n",
      "working on ...\n",
      "siTAF5L-NT.hg19 siTAF5L-NT.hg19.mapq_30.1000.mcool\n",
      "to generate siTAF5L-NT.hg19.EV.25kb.cis.bw ...\n",
      "Tokenization took: 29.61 ms\n",
      "Type conversion took: 18.56 ms\n",
      "Parser memory cleanup took: 0.01 ms\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n",
      "bedGraphToBigWig /tmp/tmpygq8qle4.bg /tmp/tmpcwndqpmh.chrom.sizes siTAF5L-NT.hg19.EV.25kb.cis.bw\n",
      "working on ...\n",
      "siTAF5L-IAA.hg19 siTAF5L-IAA.hg19.mapq_30.1000.mcool\n",
      "to generate siTAF5L-IAA.hg19.EV.25kb.cis.bw ...\n",
      "Tokenization took: 22.53 ms\n",
      "Type conversion took: 13.35 ms\n",
      "Parser memory cleanup took: 0.03 ms\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n",
      "bedGraphToBigWig /tmp/tmpwkt8mdcr.bg /tmp/tmp53brzbhf.chrom.sizes siTAF5L-IAA.hg19.EV.25kb.cis.bw\n"
     ]
    }
   ],
   "source": [
    "binsize = 25_000\n",
    "human_binsize = f\"{int(binsize/1_000)}kb\"\n",
    "# iterate over samples to calculate insulation on:\n",
    "for k,cname in ins_samples.items():\n",
    "    target_bw_file = Path(f\"{k}.EV.{human_binsize}.cis.bw\")\n",
    "    if target_bw_file.is_file():\n",
    "        print(\"already exist !\")\n",
    "        print(target_bw_file)\n",
    "        continue\n",
    "    else:\n",
    "        print(\"working on ...\")\n",
    "        print(k, cname)\n",
    "        print(f\"to generate {target_bw_file} ...\")\n",
    "        ######################################\n",
    "        ! cooltools call-compartments \\\n",
    "          --reference-track frac_gene_coverage.{human_binsize}.bedGraph \\\n",
    "          --regions hg19_arms.bed \\\n",
    "          --contact-type cis \\\n",
    "          --n-eigs 3 \\\n",
    "          -v \\\n",
    "          -o {k}.EV.{human_binsize} \\\n",
    "          --bigwig \\\n",
    "        {cname}::/resolutions/{binsize}\n",
    "\n",
    "# output of call-compartments ...\n",
    "# out_prefix + \".\" + contact_type + \".bw\"\n",
    "# out_prefix + \".\" + contact_type + \".lam.txt\"\n",
    "# out_prefix + \".\" + contact_type + \".vecs.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# binsize = 10_000\n",
    "# human_binsize = f\"{int(binsize/1_000)}kb\"\n",
    "# # input samples here\n",
    "# ev_samples = ['CkoCT442_NT_pool.hg19','CkoCT442_IAA_pool.hg19']\n",
    "# clrs = []\n",
    "# for s in ev_samples:\n",
    "#     fname = ins_samples[s]\n",
    "#     clr_uri = f\"{fname}::/resolutions/{binsize}\"\n",
    "#     clrs.append(cooler.Cooler(clr_uri))\n",
    "\n",
    "# # let's calculate gene coverage by pulling mRNA data from USCS\n",
    "# # for our hg19 genome ...\n",
    "# # this would need to be repeated for every genome, every binsize, etc\n",
    "# # probably better to download mRNA data locally to avoid redownloading ...\n",
    "# gene_cov_df = bioframe.frac_gene_coverage(\n",
    "#                 clrs[0].bins()[:],\n",
    "#                 \"hg19\"\n",
    "#                 )\n",
    "# hg19_chromsizes = bioframe.fetch_chromsizes(\"hg19\")\n",
    "\n",
    "# #saving to bigwig - whatever ...\n",
    "# for s,_r in zip(ev_samples, ev_res):\n",
    "#     # upack res, as store it:\n",
    "#     eigvals, eigvecs = _r\n",
    "#     bw_fname = f\"{s}.EV.{human_binsize}.bw\"\n",
    "#     # save compartment calls as bigwig\n",
    "#     bioframe.to_bigwig(\n",
    "#         eigvecs,\n",
    "#         hg19_chromsizes,\n",
    "#         bw_fname,\n",
    "#         value_field=\"E1\"\n",
    "#     )\n",
    "\n",
    "# # running eigdecomp ...\n",
    "# nproc = 1\n",
    "# if nproc > 1:\n",
    "#     pool = mp.Pool(nproc)\n",
    "#     map_ = pool.map\n",
    "# else:\n",
    "#     map_ = map\n",
    "\n",
    "\n",
    "# ev_res = []\n",
    "# for _clr in clrs:\n",
    "#     _gene_cov_df = copy(gene_cov_df)\n",
    "#     _res = eigdecomp.cooler_cis_eig(\n",
    "#         clr=_clr,\n",
    "#         bins=_gene_cov_df,\n",
    "#         regions=hg19_arms,\n",
    "#         n_eigs=3,\n",
    "#         phasing_track_col='coverage',\n",
    "#         balance='weight',\n",
    "#         ignore_diags=2,\n",
    "#         clip_percentile=99.9,\n",
    "#         sort_metric=None,\n",
    "#         map = map_\n",
    "#     )\n",
    "#     ev_res.append(_res)\n",
    "    \n",
    "# # plotting some stuff ...\n",
    "# from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "\n",
    "# chr1q_region = tuple(hg19_arms.iloc[1])[:-1]\n",
    "\n",
    "# for i,_r in enumerate(ev_res):\n",
    "#     a,b = _r\n",
    "#     ev1 = bioframe.select(b, chr1q_region)[\"E1\"].values\n",
    "\n",
    "#     fig = plt.figure(figsize=(15,6),constrained_layout=True)\n",
    "#     gs = fig.add_gridspec(1,3,width_ratios=[1,0.05,1])\n",
    "#     axl = fig.add_subplot(gs[0,0])\n",
    "#     cax = fig.add_subplot(gs[0,1])\n",
    "#     axr = fig.add_subplot(gs[0,2])\n",
    "\n",
    "#     hm = clrs[i].matrix(balance=True).fetch(chr1q_region)\n",
    "#     valid_mask = np.nan_to_num(hm).sum(axis=0) == 0.\n",
    "#     hm = hm[~valid_mask][:,~valid_mask]\n",
    "\n",
    "#     ccc = axl.imshow(hm,cmap=\"RdYlBu_r\",interpolation=\"nearest\",norm=LogNorm())\n",
    "#     plt.colorbar(ccc,cax=cax)\n",
    "\n",
    "#     divider = make_axes_locatable(axl)\n",
    "#     axev = divider.append_axes(\"bottom\", size=\"15%\", pad=0.08)\n",
    "\n",
    "#     axev.plot(ev1[~valid_mask],'b-')\n",
    "#     axev.axhline(0,color=\"grey\")\n",
    "#     axev.set_xlim(0,np.sum(~valid_mask))\n",
    "\n",
    "#     # color histogram\n",
    "#     _1, bin_edges, _2 = axr.hist(np.nan_to_num(np.log(hm[hm>0])).flatten(),bins=100,log=True)\n",
    "#     axr.bar(bin_edges[0],np.sum(hm==0),width=-.3,align=\"edge\",color=\"red\",alpha=0.4,label=\"zero-pixels\")\n",
    "#     axr.set_xlabel(\"heatmap_signal, excluding 0-pixels\")\n",
    "#     axr.legend(frameon=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
